magda-broken-link-sleuther/src/BrokenLinkSleuther.ts

```diff
@@ -0,0 +1,182 @@
+import Registry from '@magda/typescript-common/lib/Registry';
+import * as request from 'request';
+import * as http from 'http';
+import * as Client from 'ftp';
+import * as LRU from 'lru-cache';
+import * as URI from 'urijs';
+import retryBackoff from '@magda/typescript-common/lib/retryBackoff';
+import formatServiceError from '@magda/typescript-common/lib/formatServiceError';
+import AsyncPage, { forEachAsync } from '@magda/typescript-common/lib/AsyncPage';
+
+const aspectDefinition = {
+    id: 'source-link-status',
+    name: 'Details about the downloadURL link status of a distribution',
+    jsonSchema: require('@magda/registry-aspects/source-link-status.schema.json')
+};
+
+
+export default class BrokenLinkSleuther {
+    private registry: Registry;
+
+    constructor({ registry }: BrokenLinkSleutherOptions) {
+        this.registry = registry;
+    }
+    async run(): Promise<BrokenLinkSleutherResult> {
+        let checkedDistributions = 0;
+        let brokenLinks = 0;
+        let tooManyRequests = 0;
+
+        await this.registry.putAspectDefinition(aspectDefinition);
+
+
+        const registryPage = AsyncPage.create<{records: Array<{id: string, aspects: {'dcat-distribution-strings': {downloadURL: string}}}>, nextPageToken: string}>((previous) => {
```

Comment:
Might be a little cleaner to define an interface (or even a `type` on the line before) for this rather than putting the type inline.